tests, bug fix: DisconnectedBlockTransactions rewrite followups

ismaelsadeeq opened the pull request:
This PR is a follow-up to fix review comments and a bugfix from #28385

The PR

- Updated `DisconnectedBlockTransactions`'s `MAX_DISCONNECTED_TX_POOL` from kb to bytes.
- Moved `DisconnectedBlockTransactions` implementation code to `kernel/disconnected_transactions.cpp`.
- `AddTransactionsFromBlock` now assume duplicate transactions are not passed by asserting after inserting each transaction to `iters_by_txid`.
- Included a Bug fix: In the current master we are underestimating the memory usage of `DisconnectedBlockTransactions`.

    * When adding and subtracting `cachedInnerUsage` we call `RecursiveDynamicUsage` with `CTransaction` which invokes this [`RecursiveDynamicUsage(const CTransaction& tx)`](https://github.com/bitcoin/bitcoin/blob/6e721c923c87abdb8d99674093d08d8c17bc52c2/src/core_memusage.h#L32) version of `RecursiveDynamicUsage`, the output of that call only account for the memory usage of the inputs and outputs of the `CTransaction`, this omits the memory usage of the `CTransaction` object and the control block.
    * This PR fixes this bug by calling `RecursiveDynamicUsage` with `CTransactionRef` when adding and subtracting `cachedInnerUsage` which invokes [`RecursiveDynamicUsage(const std::shared_ptr<X>& p)`](https://github.com/bitcoin/bitcoin/blob/6e721c923c87abdb8d99674093d08d8c17bc52c2/src/core_memusage.h#L67) version of `RecursiveDynamicUsage` the output of the calculation accounts for the` CTransaction` object, the control blocks, inputs and outputs memory usage. 
    * see  [comment ](https://github.com/bitcoin/bitcoin/pull/28385#discussion_r1322948452)
- Added test for DisconnectedBlockTransactions memory limit.

glozow committed:
[test] DisconnectedBlockTransactions::DynamicMemoryUsage

ismaelsadeeq committed:
refactor: update `MAX_DISCONNECTED_TX_POOL` from kb to bytes

ismaelsadeeq committed:
move only: move implementation code to disconnected_transactions.cpp

glozow reviewed (commented):
Thanks for opening a followup! Did you want to pick up the memusage unit test too? (https://github.com/bitcoin/bitcoin/commit/9dcef47e4d52b15ffe88be7ddf5ea626121e5f0d or a new one)

ismaelsadeeq commented on src/kernel/disconnected_transactions.cpp:
Motivation is from @stickies-v  review comment https://github.com/bitcoin/bitcoin/pull/28385#pullrequestreview-1629400522
Its a better practice to have interfaces in a `.h` and implementation code in `.cpp` file, allows for more efficient compilation by preventing redundant recompilation of unchanged code, ensuring faster build times I think.

I dont think there will be need to have implementation code in the header files.
https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#source-code-organization

BrandonOdiwuor reviewed (commented):
ACK 2517862abb2c2e111c259ad2a6498e352f026c92

Great job on moving the implementation code from the header file to disconnected_transactions.cpp. This was a necessary step, especially considering how significantly the code had grown.

ismaelsadeeq committed:
assume duplicate transactions are not added to `iters_by_txid`

In `AddTransactionsToBlock` description comment we have the asuumption
that callers will never pass multiple transactions with the same txid
We are asserting to assume that does not happen.

ismaelsadeeq commented:
Added a commit to address review comment https://github.com/bitcoin/bitcoin/pull/28385#discussion_r1346498993

And rebased due to #28567  being merged for CI to be green

ismaelsadeeq commented:
> Thanks for opening a followup! Did you want to pick up the memusage unit test too? ([9dcef47](https://github.com/bitcoin/bitcoin/commit/9dcef47e4d52b15ffe88be7ddf5ea626121e5f0d) or a new one)

Added the the test.

ismaelsadeeq commented on src/test/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,90 @@
+    for (const auto& tx : block_vtx)
+        BOOST_CHECK_EQUAL(RecursiveDynamicUsage(tx), TX_USAGE);
+
+    const size_t ENTRY_USAGE_OVERESTIMATE{TX_USAGE + 8 * memusage::MallocUsage(sizeof(void*))};
```
@glozow 
the `ENTRY_USAGE_OVERESTIMATE` in https://github.com/bitcoin/bitcoin/commit/9dcef47e4d52b15ffe88be7ddf5ea626121e5f0d is too abit much and two transaction are being added to `disconnectpool` for the first  test that check that its big enough for one transaction.
So reduced the overestimate to `TX_USAGE + 8 * memusage::MallocUsage(sizeof(void*)`

glozow replied:
Makes sense to me - the numbers were a little bit off when I originally wrote that test. Thanks for fixing :+1:

glozow commented:
cc @theuni @stickies-v who made some of the suggestions being addressed and reviewed the test previously

stickies-v commented on src/test/disconnected_transactions.cpp:
Note for other reviewers: these are the memory allocation numbers I'm getting on my machine:

```
(lldb) p MAP_1
(const size_t) 32
(lldb) p MAP_100
(const size_t) 832
(lldb) p TX_USAGE
(const size_t) 640
(lldb) p ENTRY_USAGE_OVERESTIMATE
(const size_t) 896
```

This corresponds to a safety margin of 40%.

stickies-v commented on src/test/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,90 @@
+        // Only 1 transaction needed to be evicted
+        BOOST_CHECK_EQUAL(1, evicted_txns.size());
+
+        // Transactions are added from back to front and eviction is FIFO.
+        // The last transaction of block_vtx should be the first to be evicted.
+        BOOST_CHECK_EQUAL(block_vtx.back(), evicted_txns.front());
```
It's a good test to have but arguably not a `disconnectpool_memory_limits`. I think this could be a separate test case (and i also wouldn't be opposed to unit testing the rest of the interface, i.e. `AddTransactionsFromBlock`, `removeForBlock`, `take`, `clear`.

stickies-v replied:
I think it's fine to have just a single test commit here. Testing the interface just could use a separate test from the memory management test.

ismaelsadeeq replied:
Removed the test, I will like to limit this to `disconnectpool_memory_limits` test.

stickies-v replied:
If you're not adding it in a new test, I would leave it in here, it's a useful check.

stickies-v reviewed (commented):
Approach ACK 225c9c3b25d24d0f8e7730fc17ae6a7b49a2b272

Even though they're both correct and an improvement, I'd drop 6f5768d81272aeb81fd3660a26318be4e34ddd5e and 89c994acb1c12d02b072b66c7041b144fc65a6b4. For both, the fixups are arbitrary and incomplete (e.g. there are plenty of other places where we don't use list initialization in `validation.cpp`, so why are only these two being fixed? "It's a follow-up" is a bit of a wonky rationale. I'd say let's not do it, or do it consistently and ideally enforce it (which I don't think is suitable for this PR). Likewise for the clang-format fixes. Inconsistent style-only refactors are generally a hard sell.

note: I reviewed 72ebef1677a0ae61a1ca9dcde576553fb6e99fa6 with `--color-moved=zebra --color-moved-ws="allow-indentation-change"` (always useful to mention those things in PR description to help reviewers be more productive)

ismaelsadeeq commented:
Thanks for your review @stickies-v, will drop the commits as you suggested and address comments.

glozow commented on src/test/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,90 @@
+BOOST_AUTO_TEST_CASE(disconnectpool_memory_limits)
+{
+    if (sizeof(void*) != 8) {
+        BOOST_TEST_MESSAGE("Exiting disconnected_transactions memory usage test early due to unsupported arch");
```
Well the good news is it looks like it passes on 32bit, so maybe just remove this block?

ismaelsadeeq replied:
Removed

glozow commented on src/test/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,90 @@
+    // DisconnectedBlockTransactions that's just big enough for 1 transaction.
+    {
+        DisconnectedBlockTransactions disconnectpool{MAP_1 + ENTRY_USAGE_OVERESTIMATE};
+        // Just 2 transactions because the allocation depends on the number of txns passed in.
```
I meant don't pass in all 100 transactions because the memory usage won't be what we expect and iirc all of them will be evicted. The function works by first allocating 100 buckets, adding all of the transactions, and then trimming. If we add all 100, there's much more overhead and we end up not having room for anything.

stickies-v replied:
Okay, that makes sense. I didn't understand what the alternative was or what the allocation referred to exactly, so here's a suggestion to make this more explicit:
```
        // Add just 2 (and not all 100) transactions to keep the unordered map's hashtable overhead
        // to a minimum and avoid all (instead of all but 1) transactions getting evicted
```

ismaelsadeeq replied:
Added thank you.

glozow commented on src/test/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,90 @@
+    for (const auto& tx : block_vtx)
+        BOOST_CHECK_EQUAL(RecursiveDynamicUsage(tx), TX_USAGE);
+
+    const size_t ENTRY_USAGE_OVERESTIMATE{TX_USAGE + 8 * memusage::MallocUsage(sizeof(void*))};
```
The 8 pointers isn't to create an overestimation margin, it's to account for the approximate usage of the data structures (pointers in the list nodes, etc.)

glozow replied:
Maybe a comment like this would suffice?
```
// Our overall formula is unordered map overhead + usage per entry.
// Implementations may vary, but we're trying to guess the usage of data structures:
// list: per entry x (2 pointers + element itself, which is a CTransactionRef)
// unordered map: hashtable overhead + per entry x (3 pointers + value itself, which is an iterator)
// Usage per entry is thus approximately `TX_USAGE` + 6 pointers. Add another 2 pointers as a safety margin.
```

ismaelsadeeq replied:
Added the comment but I think, its word to also explain whats the pointers represent also
```
// Our overall formula is unordered map overhead + usage per entry.
// Implementations may vary, but we're trying to guess the usage of data structures:
// list: per entry x (2 pointers (prev, next) + element itself, which is a CTransactionRef)
// unordered map: hashtable overhead + per entry x (3 pointers (key, value, internal structure for book keeping) + value itself, which is an iterator)
// Usage per entry is thus approximately `TX_USAGE` + 6 pointers. Add another 2 pointers as a safety margin.
```
But not so sure if its correct for unordered map, I added as you suggested, if you think the explanation is worth it happy to repush

stickies-v replied:
In combination with [the fix suggested here](https://github.com/bitcoin/bitcoin/pull/28385#discussion_r1363659227), what do you think about this?

It makes all the assumptions explicit in code (except for the pointers which are documented in docstring). It also removes the safety margin. I think we can leave that out until/if we observe it causing issues, e.g. on certain platforms?

<details>
<summary>git diff on 8faa9800891c8b3eaef051ad9f0dc4d61475cebe</summary>

```diff
diff --git a/src/test/disconnected_transactions.cpp b/src/test/disconnected_transactions.cpp
index b488866a37..bae9850de9 100644
--- a/src/test/disconnected_transactions.cpp
+++ b/src/test/disconnected_transactions.cpp
@@ -34,10 +34,14 @@ BOOST_AUTO_TEST_CASE(disconnectpool_memory_limits)
 
     // Our overall formula is unordered map overhead + usage per entry.
     // Implementations may vary, but we're trying to guess the usage of data structures:
-    // list: per entry x (2 pointers (next pointer and prev pointer) + element itself, which is a CTransactionRef)
-    // unordered map: hashtable overhead + per entry x (3 pointers + value itself, which is an iterator)
-    // Usage per entry is thus approximately `TX_USAGE` + 6 pointers. Add another 2 pointers as a safety margin.
-    const size_t ENTRY_USAGE_OVERESTIMATE{TX_USAGE + 8 * memusage::MallocUsage(sizeof(void*))};
+    const size_t ENTRY_USAGE_OVERESTIMATE{
+        TX_USAGE
+        // list entry: 2 pointers (next pointer and prev pointer) + element itself
+        + memusage::MallocUsage((2 * sizeof(void*)) + sizeof(decltype(block_vtx)::value_type))
+        // unordered map: 1 pointer for the hashtable + key and value
+        + memusage::MallocUsage(sizeof(void*) + sizeof(decltype(temp_map)::key_type)
+                                + sizeof(decltype(temp_map)::value_type))
+    };
 
     // DisconnectedBlockTransactions that's just big enough for 1 transaction.
     {

```
<details>

ismaelsadeeq commented on src/kernel/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,86 @@
+    for (auto block_it{vtx.rbegin()}; block_it != vtx.rend(); ++block_it) {
+        auto it{queuedTx.insert(queuedTx.end(), *block_it)};
+        auto emplace_result = iters_by_txid.emplace((*block_it)->GetHash(), it);
+        assert(emplace_result.second);
```
fixed but Ive used used, `already_exist`  instead since that what the boolean represent and what cpp reference describe the boolean

ismaelsadeeq replied:
Yes you are right, my bad

ismaelsadeeq commented on src/test/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,90 @@
+        DisconnectedBlockTransactions disconnectpool{MAP_1 + ENTRY_USAGE_OVERESTIMATE};
+        // Just 2 transactions because the allocation depends on the number of txns passed in.
+        std::vector<CTransactionRef> two_txns({block_vtx.at(0), block_vtx.at(1)});
+        auto evicted_txns = disconnectpool.AddTransactionsFromBlock(two_txns);
```
fixed thank you

ismaelsadeeq commented on src/kernel/disconnected_transactions.cpp:
thanks updated

ismaelsadeeq commented on src/kernel/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,85 @@
+    std::list<CTransactionRef> ret{std::move(queuedTx)};
+    clear();
+    return ret;
+}
```
added 💯

ismaelsadeeq commented on src/kernel/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,86 @@
+#include <core_memusage.h>
+
+DisconnectedBlockTransactions::DisconnectedBlockTransactions(size_t max_mem_usage)
+    : m_max_mem_usage{max_mem_usage} {}
```
its now in the header thanks

ismaelsadeeq commented on src/kernel/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,86 @@
+
+#include <kernel/disconnected_transactions.h>
+
+#include <core_memusage.h>
```
Added Thanks

ismaelsadeeq commented:
Force pushed https://github.com/bitcoin/bitcoin/commit/225c9c3b25d24d0f8e7730fc17ae6a7b49a2b272 to 8faa980089
- Dropped the first two commits,
- Added missing include in `kernel/disconnected_transactions.cpp`,
- Updated  https://github.com/bitcoin/bitcoin/pull/28385#discussion_r1346498993 to destructure and capture the result of the emplace then assert and also.
-Move the `kernel/disconnected_transactions.cpp` constructor back to header.
- Removed the early return in [src/test/disconnected_transactions.cpp](https://github.com/bitcoin/bitcoin/pull/28530/files#diff-3cd57ed0ff682b234d9991ad80eaf492edd9f4b01f9ea822d6344fee0eac183f) since it passes on 32 bit.
- Added a comment https://github.com/bitcoin/bitcoin/pull/28530/files#diff-3cd57ed0ff682b234d9991ad80eaf492edd9f4b01f9ea822d6344fee0eac183f to explain how `ENTRY_USAGEOVERESTIMATE` is calculated
- Updated PR OP

theuni commented:
ACK the first 3 commits. I'll pass on re-reviewing the test commit and confusing myself about memory accounting again :)

stickies-v committed:
bugfix: correct DisconnectedBlockTransactions memory usage

With `queuedTx` owning the `CTransactionRef` shared ptrs, they (and
the managed objects) are entirely allocated on the heap. In
`DisconnectedBlockTransactions::DynamicMemoryUsage`, we account for
the 2 pointers that make up the shared_ptr, but not for the managed
object (CTransaction) or the control block.

Prior to this commit, by calculating the `RecursiveDynamicUsage` on
a `CTransaction` whenever modifying `cachedInnerUsage`, we account
for the dynamic usage of the `CTransaction`, i.e. the `vins` and
`vouts` vectors, but we do not account for the `CTransaction`
object itself, nor for the `CTransactionRef` control block.

This means prior to this commit, `DynamicMemoryUsage` underestimates
dynamic memory usage by not including the `CTransaction` objects and
the shared ptr control blocks.

Fix this by calculating `RecursiveDynamicUsage` on the
`CTransactionRef` instead of the `CTransaction` whenever modifying
`cachedInnerUsage`.

ismaelsadeeq commented on src/kernel/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,87 @@
+{
+    iters_by_txid.reserve(iters_by_txid.size() + vtx.size());
+    for (auto block_it{vtx.rbegin()}; block_it != vtx.rend(); ++block_it) {
+        auto it{queuedTx.insert(queuedTx.end(), *block_it)};
```
Thanks removed

stickies-v commented on src/test/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,91 @@
+    for (const auto& tx : block_vtx)
+        BOOST_CHECK_EQUAL(RecursiveDynamicUsage(tx), TX_USAGE);
+
+    // Our overall formula is unordered map overhead + usage per entry.
+    // Implementations may vary, but we're trying to guess the usage of data structures.
+    // No safety margin is added. It is assumed that this is an accurate guess. If this
+    // estimate causes issues on other platforms, then we may consider adding an additional
+    // pointer as safety margin.
```
I don't know how useful this comment is. For `ENTRY_USAGE_ESTIMATE`, we're just hardcoding how we calculate things in `memusage`, so this should always match, I think? I'd be fine removing it tbh.

ismaelsadeeq replied:
removed

stickies-v commented on src/kernel/disconnected_transactions.cpp:
```diff
@@ -51,9 +51,8 @@ size_t DisconnectedBlockTransactions::DynamicMemoryUsage() const
     for (auto block_it = vtx.rbegin(); block_it != vtx.rend(); ++block_it) {
         auto it = queuedTx.insert(queuedTx.end(), *block_it);
-        auto [_, inserted] = iters_by_txid.emplace((*block_it)->GetHash(), it);
-        assert(inserted); // callers may never pass multiple transactions with the same txid
```
I don't think you meant to remove this? It reverts f4254e209801d6a790b5f0c251c0b32154a4e3cc

ismaelsadeeq replied:
interactive rebase mistake, fixed in latest push.

stickies-v commented on src/kernel/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,90 @@
+    for (auto block_it = vtx.rbegin(); block_it != vtx.rend(); ++block_it) {
+        auto it = queuedTx.insert(queuedTx.end(), *block_it);
+        auto [_, inserted] = iters_by_txid.emplace((*block_it)->GetHash(), it);
+        assert(inserted);
```
nit: f4254e209801d6a790b5f0c251c0b32154a4e3cc has a docstring but 070f172a877a7c03fec5c9b1a032106b9a80b329 removes it:

```suggestion
        assert(inserted); // callers may never pass multiple transactions with the same txid
```

tip to avoid these issues is to run through your commits one-by-one especially after doing rebase operations etc, it's a pretty common mistake

ismaelsadeeq replied:
Thanks, will avoid.

stickies-v commented on src/test/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,94 @@
+        // list entry: 2 pointers (next pointer and prev pointer) + element itself
+        + memusage::MallocUsage((2 * sizeof(void*)) + sizeof(decltype(block_vtx)::value_type))
+        // unordered map: 1 pointer for the hashtable + key and value
+        + memusage::MallocUsage(sizeof(void*) + sizeof(decltype(temp_map)::key_type) + sizeof(decltype(temp_map)::value_type))};
```
nit: exceeds line length

ismaelsadeeq replied:
https://github.com/bitcoin-core-review-club/bitcoin/tree/pr28368/contrib/devtools#clang-format-diffpy
It was short before but contributor guide clang format e9906c105dffbb864e8c5f7680334480cdd19cd9
made the line longer?

stickies-v reviewed (approved):
ACK e9906c105dffbb864e8c5f7680334480cdd19cd9

Just 2 nits, happy to quickly re-ack since no other acks yet.

edit: PR description needs updating, would change title to `tests, bugfix` and add the bugfix in the description, as well as the "assume duplicate transactions are not added to `iters_by_txid`" commit

stickies-v commented on src/kernel/disconnected_transactions.cpp:
```diff
@@ -0,0 +1,90 @@
+        if (iter != iters_by_txid.end()) {
+            auto list_iter = iter->second;
+            iters_by_txid.erase(iter);
+            cachedInnerUsage -= RecursiveDynamicUsage(**list_iter);
```
Hmm this didn't get cherry-picked properly from https://github.com/stickies-v/bitcoin/commit/87f3f28b3acfc329eaed0b0cce6ff589271bfefd#diff-daf5057c112d6ec1338dd535558cd0f4e21e9a3e0241b610d88068fe2baba362R113 , probably because of rebase?

ismaelsadeeq commented:
> edit: PR description needs updating, would change title to `tests, bugfix` and add the bugfix in the description, as well as the "assume duplicate transactions are not added to `iters_by_txid`" commit

Thank you @stickies-v , I updated the PR description and address review comments.

stickies-v reviewed (approved):
ACK 9b3da70bd06b45482e7211aa95637a72bd115553 - nice work!

BrandonOdiwuor reviewed (approved):
re ACK 9b3da70bd06b45482e7211aa95637a72bd115553

glozow commented on src/kernel/disconnected_transactions.cpp:
```diff
@@ -34,7 +34,7 @@ std::vector<CTransactionRef> DisconnectedBlockTransactions::LimitMemoryUsage()
     while (!queuedTx.empty() && DynamicMemoryUsage() > m_max_mem_usage) {
         evicted.emplace_back(queuedTx.front());
-        cachedInnerUsage -= RecursiveDynamicUsage(*queuedTx.front());
+        cachedInnerUsage -= RecursiveDynamicUsage(queuedTx.front());
```
For future reference for b2d04479647af64ad7cf5ebfb6175251b2f6b72e 

Imagining `DynamicMemoryUsage` for a `DisconnectedBlockTransactions` with 1 transaction (ignoring the `iters_by_txid` part since that doesn't change),

before
```
DynamicUsage(queuedTx) + cachedInnerUsage
=MallocUsage(sizeof(list_node<CTransactionRef>) + cachedInnerUsage
=MallocUsage(sizeof(4 pointers) + RecursiveDynamicUsage(CTransaction)
=MallocUsage(sizeof(4 pointers) + DynamicUsage(vin) + DynamicUsage(vout) + sum([RecursiveDynamicUsage(input) for input in vin]) + sum([RecursiveDynamicUsage(output) for output in vout])
```
after
```
DynamicUsage(queuedTx) + cachedInnerUsage
=MallocUsage(sizeof(list_node<CTransactionRef>) + cachedInnerUsage
=MallocUsage(sizeof(4 pointers) + RecursiveDynamicUsage(CTransactionRef)
=MallocUsage(sizeof(4 pointers) + DynamicUsage(CTransactionRef) + RecursiveDynamicUsage(CTransaction)
=MallocUsage(sizeof(4 pointers) + MallocUsage(sizeof(CTransaction)) + MallocUsage(sizeof(stl_shared_counter)) + RecursiveDynamicUsage(CTransaction)
=MallocUsage(sizeof(4 pointers) + MallocUsage(sizeof(CTransaction)) + MallocUsage(sizeof(stl_shared_counter)) + DynamicUsage(vin) + DynamicUsage(vout) + sum([RecursiveDynamicUsage(input) for input in vin]) + sum([RecursiveDynamicUsage(output) for output in vout])
```

So we account for
- each list node, which is 2 pointers + a shared pointer object (2 pointers)
- each shared pointer's dynamically allocated memory i.e. the `CTransaction` object + its always dynamically allocated stuff (the vectors) + the control block

glozow reviewed (commented):
ACK 9b3da70bd06b45482e7211aa95637a72bd115553
